pkg/controller/clusterpool/clusterpool_controller.go

```diff
@@ -270,48 +283,22 @@ func (r *ReconcileClusterPool) Reconcile(ctx context.Context, request reconcile.
 		return reconcile.Result{}, nil
 	}
 
-	// Find all ClusterDeployments from this pool:
-	claimedCDs, unClaimedCDs, err := r.getAllClusterDeploymentsForPool(clp, logger)
+	cds, err := getAllClusterDeploymentsForPool(r.Client, clp, logger)
 	if err != nil {
 		return reconcile.Result{}, err
 	}
 
-	var toRemoveClaimedCDs []*hivev1.ClusterDeployment
-	numberOfDeletingClaimedCDs := 0
-	for _, cd := range claimedCDs {
-		toRemove := controllerutils.IsClaimedClusterMarkedForRemoval(cd)
-		switch {
-		case cd.DeletionTimestamp != nil:
-			numberOfDeletingClaimedCDs++
-		case toRemove:
-			toRemoveClaimedCDs = append(toRemoveClaimedCDs, cd)
-		}
-	}
-
-	var installingCDs []*hivev1.ClusterDeployment
-	var readyCDs []*hivev1.ClusterDeployment
-	numberOfDeletingCDs := 0
-	for _, cd := range unClaimedCDs {
-		switch {
-		case cd.DeletionTimestamp != nil:
-			numberOfDeletingCDs++
-		case !cd.Spec.Installed:
-			installingCDs = append(installingCDs, cd)
-		default:
-			readyCDs = append(readyCDs, cd)
-		}
+	claims, err := getAllClaimsForPool(r.Client, clp, logger)
+	if err != nil {
+		return reconcile.Result{}, err
 	}
 
-	logger.WithFields(log.Fields{
-		"installing": len(installingCDs),
-		"deleting":   numberOfDeletingCDs,
-		"total":      len(unClaimedCDs),
-		"ready":      len(readyCDs),
-	}).Debug("found clusters for ClusterPool")
+	claims.SyncCDAssignments(r.Client, cds, logger)
+	cds.SyncClaimAssignments(r.Client, claims, logger)
```

Comment:
From what i can see, if the claim user, where to create a claim with a spec.namespace set to an cd in the pool, this would sync the cd to the claim even when the controller had not made that choice. That seems problematic, don't know if the previously it behaved like that.

I personally feel the controller should be the one in control for safety.
- since only the controller has access to the cds, we should assign cds to claim for source of truth
- sync mis match only when the cd says it's assigned to the claim.

Reply:
> don't know if the previously it behaved like that.

It did, for better or worse. Previously the clusterpool_controller would have [ignored](https://github.com/openshift/hive/blob/34d8b527b80acfb55701034adfff9caf978e3326/pkg/controller/clusterpool/clusterpool_controller.go#L920-L923) such a claim, and the clusterclaim_controller would have
- [assigned the CD](https://github.com/openshift/hive/blob/34d8b527b80acfb55701034adfff9caf978e3326/pkg/controller/clusterclaim/clusterclaim_controller.go#L431) if it was unassigned;
- [accepted the CD assignment](https://github.com/openshift/hive/blob/34d8b527b80acfb55701034adfff9caf978e3326/pkg/controller/clusterclaim/clusterclaim_controller.go#L268) if it matched; or
- [flagged a conflict](https://github.com/openshift/hive/blob/34d8b527b80acfb55701034adfff9caf978e3326/pkg/controller/clusterclaim/clusterclaim_controller.go#L495) if it didn't.

If that's not a desired behavior, I suggest we address it separately. (I can see it being a "feature": a way to allow the user to claim a specific CD from the pool, which may dovetail into the whole HIVE-1058 issue.)

What is new here is that we'll sync the _other_ way. I'm not sure what scenario would cause a CD to be assigned to a claim that's not assigned to the CD. It wouldn't be a partial failure in this code, because we [update the claim first](https://github.com/openshift/hive/pull/1474/files#diff-bbf4d3cb24b8cac87951b84ca973747902c00d4a9d662648c8e760a37569581dR511-R519) -- but I suppose again if someone manually edited the claim...

Reply:
> It did, for better or worse.

nothing we can do then here. will have to accept the current behavior though it sounds wrong.